### ISSUE TITLE
pihole-FTL: Operation not permitted

### ISSUE THREAD

**Issue by olafdietsche** [pi-hole/docker-pi-hole#975]
### Versions
  Pi-hole version is v5.6 (Latest: v5.6)
  AdminLTE version is v5.8 (Latest: v5.8)
/opt/pihole/version.sh: line 19: /usr/bin/pihole-FTL: Operation not permitted
/opt/pihole/version.sh: line 122: /usr/bin/pihole-FTL: Operation not permitted
  Latest FTL version is v5.11

### Platform
- OS and version:  Ubuntu focal
- Platform:  Docker

### Expected behavior

No error message, pihole-FTL running

### Actual behavior / bug

pihole-FTL aborts with "Operation not Permitted"

### Steps to reproduce

Build Docker image with attached files, Dockerfile, setupVars.conf, pi-hole-installer.sh
The base image was created with

    debootstrap --variant=minbase focal /path/to/target

pi-hole-installer.sh was downloaded with

    curl -sSL https://install.pi-hole.net >pi-hole-installer.sh

build*.log is the log output of the Docker build.

Attachements are renamed to *.txt, because otherwise the upload dialog would not show the files.

[Dockerfile.txt](https://github.com/pi-hole/pi-hole/files/7574450/Dockerfile.txt)
[setupVars.conf.txt](https://github.com/pi-hole/pi-hole/files/7574451/setupVars.conf.txt)
[pi-hole-installer.sh.txt](https://github.com/pi-hole/pi-hole/files/7574452/pi-hole-installer.sh.txt)
[build-2021-11-20-12.01.21.log](https://github.com/pi-hole/pi-hole/files/7574453/build-2021-11-20-12.01.21.log)

## Debug Token

The debug token was created with

    docker run --rm --interactive --tty --name test pihole pihole -d

- URL:  https://tricorder.pi-hole.net/mBQRFys0/

**Comment by PromoFaux** [comment]
> /opt/pihole/version.sh: line 19: /usr/bin/pihole-FTL: Operation not permitted
> /opt/pihole/version.sh: line 122: /usr/bin/pihole-FTL: Operation not permitted

I recall seeing this at some point in the past when I was working on improvements to our own build, I don't recall the reason for it in the end (a lot has changed since November!)

See our Dockerfile here: https://github.com/pi-hole/docker-pi-hole/blob/master/Dockerfile

And the install script it runs: https://github.com/pi-hole/docker-pi-hole/blob/master/install.sh

Just out of curiosity, is there a reason you're trying to build your own (not that there is anything wrong with that!) rather than use our image?

**Comment by olafdietsche** [comment]
The reason is mainly understanding what happens and minimalism.
I know about pi-hole-docker repository, but I wanted to run without s6 or docker-compose, so I tried my own.

**Comment by PromoFaux** [comment]
Yep, that's fair enough, there are no wrong answers. To be honest - I'm not sure it's even possible without an init layer (which is what s6 is in my understanding) - but most of my experience with docker has just been in tweaking this particular image and seeing what I can get away with 😏

**Comment by olafdietsche** [comment]
Ok, I am like-minded ;-)

If time permits, I will try your Docker image and see, how to tweak this.
Although, I have no experience with docker-compose yet.

**Comment by edgd1er** [comment]
I'm also building my own pihole image to test or add some features ( web logs to stdout, UID/GID, plain debian base image, dockerfile optimisation/simplification, ...)
I had to change in the dockerfile bash -ex ./install.sh to bash -x ./install.sh as I had the same error during the install process.

L77:  ... pihole-FTL branch
L91: .... pihole-FTL version

my forked pi-hole (same error as olaf)
```
2022-01-09T07:57:08.8501738Z #24 71.92   [✗] DNS service is NOT running
2022-01-09T07:57:08.8502352Z #24 71.92 + /opt/pihole/updatecheck.sh
2022-01-09T07:57:08.8503589Z #24 71.94 /opt/pihole/updatecheck.sh: line 77: /usr/bin/pihole-FTL: Operation not permitted
2022-01-09T07:57:08.8504873Z #24 72.11 /opt/pihole/updatecheck.sh: line 91: /usr/bin/pihole-FTL: Operation not permitted
2022-01-09T07:57:08.8505770Z #24 72.11 + /opt/pihole/updatecheck.sh x remote
``` 

upstream pi-hole ( no error)
```
2022-01-08T13:44:28.5879239Z #22 22.66   [✗] DNS service is NOT running
2022-01-08T13:44:28.7379008Z #22 22.67 + /opt/***/updatecheck.sh
2022-01-08T13:44:28.8884740Z #22 22.82 + /opt/***/updatecheck.sh x remote
```

Once the image is build, in the running container commands are ok:
```
root@pihole:/# pihole-FTL branch
master
root@pihole:/# pihole-FTL version
v5.13
```

not really a problem ate the moment, but I'm looking into it to better understand the install process.

**Comment by olafdietsche** [comment]
So it might be the build and has nothing to do with Docker itself.
Thank you for the hint.

**Comment by edgd1er** [comment]
I've inserted some commands to check pihole-FTL into pihole install script.
Even after a pihole -r, during build time, pihole-FTL cannot be executed, even to display help messages.
I will look if the docker builder is not set in a specific way.

```
#12 19.56   [✗] DNS service is NOT running
#12 19.57 + ls -al /usr/bin/pihole-FTL
#12 19.57 -rwxr-xr-x 1 root root 15333192 Jan 10 08:03 /usr/bin/pihole-FTL
#12 19.57 + whoami
#12 19.57 root
#12 19.57 + /usr/bin/pihole-FTL debug test
#12 19.57 /tmp/ph_install.sh: line 2606: /usr/bin/pihole-FTL: Operation not permitted
#12 19.57 + /usr/local/bin/pihole -r
.......
#12 21.63 + /usr/bin/pihole-FTL --help
#12 21.63 /usr/local/bin/install.sh: line 59: /usr/bin/pihole-FTL: Operation not permitted
#12 21.63 + /usr/bin/pihole-FTL debug test
#12 21.63 /usr/local/bin/install.sh: line 60: /usr/bin/pihole-FTL: Operation not permitted
```

**Comment by edgd1er** [comment]
I had a look at the buildkit and even if my download buildkit 0.7.3 , it is later upgraded to 0.9.3.

Gitlab buildkit version is 0.9.3. , prior versions introduced [security tweakings](https://github.com/moby/buildkit/issues/950).

The buildx builder has to be created in its latest version with specific flags:
```
docker buildx create --driver-opt image=moby/buildkit:master  --use --name insecure-builder --buildkitd-flags '--allow-insecure-entitlement security.insecure'
```
the job has to be run with `docker buildx build --allow security.insecure .....`

well, all this is theoretical. It didnt't change anything about `operation not permitted`error.

So no clue for the moment; why original job is ok, but forked one has errors.

**Comment by edgd1er** [comment]
Still stuck with that issue: 

I noticed something strange. the file has the proper hash and rights (I still get the error at build time). at runtime, renaming the does not fix the issue.
Copying the file, does fix the issue. I'm not sure what is the proble.


```
root@366d4a8d97dc:/usr/bin# pihole-FTL version
bash: /usr/bin/pihole-FTL: Operation not permitted
root@366d4a8d97dc:/usr/bin# strace pihole-FTL version
execve("/usr/bin/pihole-FTL", ["pihole-FTL", "version"], 0x7edc6a64 /* 37 vars */) = -1 EPERM (Operation not permitted)
strace: exec: Operation not permitted
+++ exited with 1 +++
root@366d4a8d97dc:/usr/bin# mv pihole-FTL a
root@366d4a8d97dc:/usr/bin# strace a
execve("/usr/bin/a", ["a"], 0x7e94da70 /* 37 vars */) = -1 EPERM (Operation not permitted)
strace: exec: Operation not permitted
+++ exited with 1 +++
root@366d4a8d97dc:/usr/bin# cp a pihole-FTL
root@366d4a8d97dc:/usr/bin# pihole-FTL version
v5.14
```

**Comment by edgd1er** [comment]
Seems Raspberry OS (11.0) has a special behaviour.
I did add the fix_capabilities at build time. It doesn't help.
running manually the fix_capabilities fix the problem when run in console, not when running the service.

```
root@47de57478eb3:/usr/bin# pihole-FTL
bash: /usr/bin/pihole-FTL: Operation not permitted
root@47de57478eb3:/usr/bin# getcap pihole-FTL 
pihole-FTL cap_chown,cap_net_bind_service,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_sys_nice=eip
root@47de57478eb3:/usr/bin# source /bash_functions.sh 
root@47de57478eb3:/usr/bin# fix_capabilities 
root@47de57478eb3:/usr/bin# pihole-FTL version
v5.14
root@47de57478eb3:/usr/bin# getcap pihole-FTL 
pihole-FTL cap_chown,cap_net_bind_service,cap_net_admin,cap_net_raw,cap_sys_nice=ei
```

```
root@47de57478eb3:/usr/bin# pihole-FTL version
bash: /usr/bin/pihole-FTL: Operation not permitted
root@47de57478eb3:/usr/bin# setcap = pihole-FTL 
root@47de57478eb3:/usr/bin# pihole-FTL version
v5.14
root@47de57478eb3:/usr/bin# service pihole-FTL restart
Not running

sh: 1: /usr/bin/pihole-FTL: Operation not permitted

root@47de57478eb3:/usr/bin# pihole-FTL version
bash: /usr/bin/pihole-FTL: Operation not permitted
```

**Comment by edgd1er** [comment]
Well, the problem seems to be with my raspberry pi3 docker builder.
I created a new builder but that didn't change the result.
When building for linux/arm/v7 with buildx on a amd64 and pushing/pulling the image, no more permission denied problem at runtime. 

I ended up in reinstalling the host. (rpi3), the error is still present at build time (on the rpi3), not anymore at run time.
it seems that raspbian docker is not as resilient as amd64's version.

I guess this issue may be closed.
